### PR TITLE
Support for Google Cloud Storage

### PR DESCRIPTION
**Pull request by ljagiello**
Similar to Amazon S3, Google offers Cloud Storage. My patch adds support for GCS and allows to store backups in S3 and GCS if both backends are configured.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -6,14 +6,18 @@ deps:
...
 	go get -u -v github.com/aws/aws-sdk-go
-	go get -u -v  github.com/hashicorp/consul
-	go get -u -v  github.com/Pallinder/go-randomdata
-	go get -u -v  github.com/mitchellh/cli
-	go get -u -v  github.com/golang/lint/golint
+	go get -u -v github.com/hashicorp/consul
+	go get -u -v github.com/Pallinder/go-randomdata
+	go get -u -v github.com/mitchellh/cli
+	go get -u -v github.com/golang/lint/golint
+	go get -u -v golang.org/x/net/context
```
Comment (sean-) [inline Makefile]: I don't think this will be required when Go 1.7 comes out.

File: backup/backup.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ import (
 	"github.com/pshima/consul-snapshot/consul"
 	"github.com/pshima/consul-snapshot/crypt"
 	"github.com/pshima/consul-snapshot/health"
+	"golang.org/x/net/context"
```
Comment (sean-) [inline backup/backup.go]: Can be removed once 1.7 comes out.

File: backup/backup.go [hunk 3]
```diff
@@ -311,18 +311,37 @@ func (b *Backup) writeBackupRemote() {
...
 	}
 
-	// Create the params to pass into the actual uploader
-	params := &s3manager.UploadInput{
-		Bucket: &b.Config.S3Bucket,
-		Key:    &b.RemoteFilePath,
-		Body:   bytes.NewReader(localFileContents),
-	}
+	if len(b.Config.S3Bucket) > 1 {
+		s3Conn := session.New(&aws.Config{Region: aws.String(string(b.Config.S3Region))})
```
Comment (sean-) [inline backup/backup.go]: Can this be made into a dedicated function like `writeBackupRemoteS3()` and below a similar function called `writeBackupRemoteGoogleStorage()`?  Then the logic becomes very easy to follow (and testable!).

File: restore/restore.go [hunk 4]
```diff
@@ -125,17 +125,38 @@ func getRemoteBackup(r *Restore, conf *config.Config) {
 		log.Fatalf("[ERR] Unable to create local restore temp file!: %v", err)
 	}
 
-	// Create the params to pass into the actual downloader
-	params := &s3.GetObjectInput{
-		Bucket: &conf.S3Bucket,
-		Key:    &r.RestorePath,
-	}
+	if len(string(conf.GCSBucket)) < 1 {
```
Comment (sean-) [inline restore/restore.go]: Similar to the above comments, can this be moved into its own dedicated function, like `getRemoteBackupS3()` and `getRemoteBackupGoogleStorage()` ?

File: Makefile (hunk 1 above)
Comment (ljagiello) [inline Makefile]: Hi @sean- thanks for comments, should I leave that file as it is right now and skip google storage stuff?

File: backup/backup.go (hunk 3 above)
Comment (ljagiello) [inline backup/backup.go]: Sure, I will fix that later today.

File: backup/backup.go [hunk 5]
```diff
@@ -2,6 +2,7 @@ package backup
 
 import (
 	"bytes"
+  "context"
```
Comment (sean-) [inline backup/backup.go]: `go fmt` or [`goimports`](https://godoc.org/golang.org/x/tools/cmd/goimports)

File: restore/restore.go [hunk 6]
```diff
@@ -3,6 +3,7 @@ package restore
 import (
 	"bytes"
 	"compress/gzip"
+  "context"
```
Comment (sean-) [inline restore/restore.go]: Ditto `go fmt` or `goimports`.

**Comment by ljagiello** [comment]
Looks like upstream golang branch is broken (https://github.com/golang/go/issues/16741) and travis fail on that.

**Comment by ljagiello** [comment]
Hi @sean- any other comments?

**Comment by pshima** [comment]
I don't have a google account yet to try this out but looks good to me - waiting for further input from @sean- but will cut another release with this functionality after it is merged and tested.

Thanks for this fantastic feature addition @ljagiello!

**Comment by player1-fleetsmithsupport** [comment]
Hi @pshima any luck maybe?

**Comment by pshima** [comment]
Hi there, haven't heard from @sean- and do not have any google cloud credentials to test it.  I'll try and get that setup, test it, and merge this.

**Comment by sean-** [comment]
This is actually on my list to examine. I've read through the code, just need to test.

**Comment by player1-fleetsmithsupport** [comment]
Hi @pshima @sean- any luck with test/merge?

**Comment by pshima** [comment]
Well after nearly a year I decided to just merge this and will sign up for a GCE account to test.

**Pull request merged**